### ISSUE TITLE
Need a realistic sample

### ISSUE THREAD

**Issue by stefandevo** [jonataslaw/get#62]
Hi, this library has great potential, but I am missing a realistic sample.
Would like to see how this library is used with a view, viewcontroller/viewmodel, service, etc. I fully understand the navigation benefits, but the Simple State Manager part is a bit unclear in relation to provider, ChangeNotifier, etc.
Would it be possible to explain this a bit more with a real life example?

**Comment by jonataslaw** [comment]
According to official Flutter documentation:
"ChangeNotifier is optimized for small numbers (one or two) of listeners."
https://api.flutter.dev/flutter/foundation/ChangeNotifier-class.html

I have worked closely with Flutter for the past two years, and I understand the reason for ChangeNotifier is not adopted by large cases.
ChangeNotifier registers listeners to a class, and goes through the entire tree to update a single widget. ChangeNotifier is expensive, has a high consumption of ram, and succumbs to medium and large projects, and must be used for one or two listeners (according to official documentation).
Provider is very easy, but it uses ChangeNotifier and InheritedWidget, and for me there are two problems. In this library, we exterminate everything related to InheritedWidget, because it requires a context to access any component. Obviously this has advantages for certain types of structures, but for others, it is horrible.
ChangeNotifier is self explanatory, the Flutter team itself has reiterated many, many times that it works to insert a shopping cart into the application, or something that depends on the global scope, only once, or at most two. You cannot manage the status of an application with ChangeNotifier if you have 50 stateful widgets.

This new state manager works differently from any state manager released so far.
With him:
-You should never use StatefulWidget. All of your widgets will be stateless, except widgets that have been. This is a moderate performance gain depending on the size of your application.
When you create a Stateful class, all widgets, including your AppBar, have a state. This is illogical, it is a waste of resources, and this is fixed in this state manager.
-All controllers are inserted into other controllers, you must not insert ANY business logic within the UI. Well, for me, it is clearer to employ a standard like MVC with this lib. Controllers need a StatefulWidget, often you need to initialize them or add a listener to them that depend on initState. This is also corrected in that library. GetBuilder has the property "initState" and "dispose", in which you can call functions of the controller class to be initialized when the widget is built and removed when the widget is removed from memory. Instead of using FutureBuilder, you can listen to a change in the API for the reconstruction of a listview, where you will insert the API requests in the class repository, request the data and save it in your controller class, and access it in your UI. To fire the "fetchData" trigger you can use GetBuilder's initState for this.
-It is the lightest and simplest application state manager available (assuming that setState is an Ephemeral State Manager). It works as follows:
Each GetBuilder has a state. the ID of that state is saved in a Map, and updated when necessary. It all comes down to that. There are only 95 lines of code (against over 4k of Provider lines, over 8k of BLoC lines, more than 8k of MobX lines).
Obviously, it has been released now, there are many improvements to be made. The reactive model (which works with autodisposable Streams) is yet to come in the next big update for even bigger applications, but if someone is looking for a simple, lightweight and good performance state management, it is the ideal option. It is not ideal for all cases, it does not use Streams like BLoC, but for certain cases, it is perfect.

Finally, I will make a more robust example application, with all the options available from Get, and showing what it would be like to structure a project with it. I know that things may not be clear, but my time is scarce, I am the sole maintainer of this lib, and I have guaranteed that it is more stable than Flutter's own navigation (while the master and dev were broken with routes named by 1 month, Get remained stable, adding the hotfix on the same day of the break, and before the analysis of the Flutter team).

**Comment by stefandevo** [comment]
Thanks for this extended explanation. A real life sample would do good! (a login screen, a list-view, and add/modify of an item, that use a -dummy- web api and store some data).

As I understand it now, you don't use something like get_it then for DI. You just use something like `static Controller get to => Get.find();` to get an instance of a controller anywhere in your code. 

Do you use the same for non-controllers? I mean, a Service (business logic) that needs a Repository. Is this repository also in the Get.find(), or do you use get_it ?

Furthermore, is the system easily extendable? Meaning, with different type of dialog boxes, or snackbars?

**Comment by jonataslaw** [comment]
> Obrigado por esta explicação estendida. Uma amostra da vida real faria bem! (uma tela de login, uma exibição de lista e adicionar / modificar um item, que usa uma API -dummy- web e armazena alguns dados).
> 
> Pelo que entendi agora, você não usa algo como get_it então para DI. Você apenas usa algo como `static Controller get to => Get.find();`para obter uma instância de um controlador em qualquer lugar do seu código.
> 
> Você usa o mesmo para não-controladores? Quero dizer, um Serviço (lógica de negócios) que precisa de um Repositório. Este repositório também está no Get.find () ou você usa o get_it?
> 
> Além disso, o sistema é facilmente extensível? Significado, com diferentes tipos de caixas de diálogo ou barras de lanches?
> 
> E, para ficar claro, você envolve todos os widgets da árvore que deseja atualizar de um Controller em torno do `GetBuilder`? Então, por exemplo, você tem uma tela com 4 widgets que obtêm seus valores de 4 propriedades, agrupa todos os 4 widgets com `GetBuilder`?


In fact Get uses an instance manager (just like Get_it)
Could you do something like:
`Get.put(Controller()); 
`
And use:

```dart
GetBuilder<Controller>( 
builder:(_)=> Text(Get.find<Controller>().counter)
```
I just don't recommend this approach (similar to Get_it) because it is bad.
Not everything you see in 'examples' works with large applications, Get is used with applications with more than 2 million downloads, so the recommended approach is one that does not have side effects (I'm not saying that Get_it cannot be used in large applications, because it is probably used, but that in order for it to work properly, the programmer needed to create his own dependency creation and destruction logic).
You know that Get_it registers a singleton that allocates memory indefinitely, and you need to manually remove it from memory when you want to free up resources.
Using GetBuilder's `init: Controller()`, you will register this dependency, and as soon as it is no longer useful, it will be discarded, and you need not reason "when" it will be useful, Get itself does it for you if it detects that you don't use it anywhere else in your code.
If you use GetBuilder in one place, as soon as the GetBuider that started the controller comes out of memory, Get will automatically remove all its controllers relationed from memory. If you use 30 widgets, as soon as no screen uses the controller, it will remove this dependency from memory.
As I said, you can use this package as you would use Provider/Get_it, but that means you will have the same performance problems as using Provider/Get_it.

Screen A => Screen B (has GetBuilder) => Screen C (has GetBuilder)
You navigate to Screen A = Controller is not in memory 
You navigate to Screen B = Controller is initialized and it's in memory 
You navigate to Screen C = Get realizes that you use GetBuilder with the same controller, and will send data from controller used on screen B. 
You leave the screen C, Get remain with the Controller in memory because it's used on B. 
You back to screen A, Get remove Controller from memory because it is not necessary

That's why I recommend this approach, you don't have to worry about anything but programming, your app is optimized entirely by Get.

Actually Get uses its own dependency manager (which works like Get_it, but is much leaner).
If you read the documentation you will see the Get.put, Get.find, and Get.delete methods.
This would be just like Get_it registerSingleton, and we use this API to manage dependencies.
The difference is that we have another API that locates the controllers, and that it is possible to find classes using a static method inside it using Get.find();

You can use Get.put() and Get.find() for repositories, blocs, and controllers. We don't use get_it, we have our own dependency manager (Get Instance Manager) which is much leaner and smart.

As for your question about grouping widgets, you should check what is best for your App's performance. 

As for your question about grouping widgets, you should check what is best for your App's performance. Let's say you have a column that displays Video, Maps or any widget that uses PlatFormView (highly heavy widget) or animations. 

If you don't want these widgets to be rebuilt, you should do this:

```dart
Column( children[
HeavyWidget()
GetBuilder<Controller>( 
builder:(_)=> Text(Get.find<Controller>().counter)
HeavyWidget()
HeavyWidget()
GetBuilder<Controller>( 
builder:(_)=> Text(Get.find<Controller>().counter)
GetBuilder<Controller>( 
builder:(_)=> Text(Get.find<Controller>().counter)

```

However, let's say that your controller changes a text and an icon in your app, you can do it smoothly:

```dart
GetBuilder<Controller>( 
builder:(_)=> Column(children: [
Text(_.mytext), 
Icon(_.icons),
Text(_.mytext), 
Icon(_.icons),
```
You can also componentize parts of your screen. Say you want to change the state of a map (which is heavy) and a video (which is also heavy) at different times, with different logic, you could create a "components" folder, insert a MapController and VideoController together with your video widget and its map widget, You will have access to both main screen controllers if you need to. But use this only if necessary, as a rule, you can use one controller per screen, but obviously, if you have non-homogeneous state changes, for me, it makes more sense to componentize the state-changing widget, but it's my project structure , you can do it as you see fit.

Note - If you create components for your screen in the same way, you should not worry about the layout of your controller, as soon as the screen that uses these components comes out of the foreground, it will be removed from memory as well (unless you navigate to a next route that also uses this controller), Get uses the GetObserver API (which manages and observes the change of routes) and integrates fully with Get instance manager, and Get state manager, therefore, however much you add Provider, Get_it, and other plugins, you will never have this integration that Get has.
If a route is sent or popped, Get notifies the dependency manager, and it notifies the state manager. The state manager, in turn, checks if it has a GetBuilder still using the controller, if not, it notifies the instance manager to remove it from memory, and all this happens in the background.

There is basically no open issue (other than this issue), because if you use Get as the documentation, you will not face any problems with your app. Get was an internal package for my startup and was used on a large scale for about 1 year, when I decided to make it open source to help other devs looking for high performance, high productivity, and stability solutions.
Obviously, you can use it in another way, and even invent your own dependency management logic, but following the pattern, you can program without fear of launching your app and find future bugs related to navigation, instance management or dependencies. Any other way, it may or may not have side effects.

**Comment by stefandevo** [comment]
Thanks again for the extensive reply. I got the philosophy behind this now; thanks.
So basically you do use the Get instance manager for registering your blocs/repositories also. Cool. So no more get_it, no more provider, changenotifier, etc.
A last question on testing. How can I register the bloc's abstract class with a specific implementation for the app and for testing?

**Comment by jonataslaw** [comment]
You can't register the bloc's abstract class, cuz abstract class hasn't instances.

And by the way, I changed the example application to a realistic application. If you have another problem, do not hesitate to open another issue.
Closing this here, but I will still answer if I have any other questions not related to the issue.
